### PR TITLE
Add error handler

### PR DESCRIPTION
**Pull request by syngenta-aleksei-zolotykh**
Sometimes we need to handle errors that happened while request proceed. We need it, for example, for sending reports to the monitoring system. 
```
            this.router = new Router({
              
                beforeAll: () => {
                   throw error; // <= error may occurs in any place till the process is not
                },
                onError: (error, response) => sendDataToMonitoringService(error)
            });

```

### CODE REVIEW

File: src/apigateway/router.js [hunk 1]
```diff
@@ -43,6 +44,9 @@ class Router {
                 response: this._errors
             });
         }
+        if (typeof this._onError === 'function') {
+            this._onError(error, this._errors);
+        }
```
Comment (paulcruse-syn) [inline src/apigateway/router.js:49]: The function signature should be the same as all the other function contexts that we use. If you look at the `_beforeAll` and `_afterAll`, we pass:

```javascript
await this._beforeAll(request, response, endpoint.requirements);
await this._afterAll(request, response, endpoint.requirements);
```
Obviously, we don't need the requirements on this, but we should still be passing in request and response objects, followed by an additional error object.
 
We should be doing something like:

```suggestion
        if (typeof this._onError === 'function') {
            this._onError(request, response, error);
        }
```

You will need to mutate the response object to be able to get a response out to the API end user. 

Lastly, I don't think we should run both this and if statements above L33 `if (!this._errors.hasErrors)` it should be one or the other, we don't want mutate the error object with our standard response if the user is writing their own error handler, this whole handle method error needs to be refactored, something like:


```javascript
const request = new RequestClient(this._event);
const response = new ResponseClient();
if (typeof this._onError === 'function') {
    this._onError(request, response, error);
} else if (!this._errors.hasErrors) {
    this._errors.code = 500;
    this._errors.setError('server', 'internal server error');
}
if (!process.env.unittest) {
            this._logger.error({
                error_messsage: error.message,
                error_stack: error.stack instanceof String ? error.stack.split('\n') : error,
                event: this._event,
                request: request.request,
                response: this._errors
      });
}
```

**Review by paulcruse-syn — CHANGES_REQUESTED** [review]
Looks good! Great idea! 

We need to refactor the `__handleError` method so that we are either using our standard error response or running the users custom error handler.

File: src/apigateway/router.js (hunk 1 above)
Comment (syngenta-aleksei-zolotykh) [reply src/apigateway/router.js]: Thanks for the so detailed review. I've fixed all issues.

**Review by paulcruse-syn — APPROVED** [review]

**Comment by paulcruse-syn** [comment]
This is now being released as [1.0.12](https://github.com/syngenta-digital/package-node-alc/releases/tag/1.0.12)

**Pull request merged**